File: web/src/lib/ignore.ts
/**
 * What the panel's folder picker leaves out of a folder before sending it.
 *
 * The same rules as `skifity up` (internal/cli/ignore.go), and held to the same
 * cases: internal/cli/testdata/ignore-cases.json is read by the Go test and by
 * web/scripts/check-ignore.mjs, which runs this file. Two implementations of
 * one rule stay one rule only that way.
 *
 * A folder somebody picks has node_modules in it, a .next cache, the .env with
 * their real keys. .gitignore already draws that line, so it is read the way
 * git reads it, in every folder, and .skifityignore adds to it.
 *
 * No imports, so Node can load this file directly for the check.
 */

/** Never sent, whatever an ignore file says. */
const alwaysLeftOut = new Set([
  ".git",
  ".hg",
  ".svn",
  "node_modules",
  ".DS_Store",
  "Thumbs.db",
  "skifity.toml",
])

/** Left out as if a .gitignore said so, so a "!" line can bring one back. */
const leftOutUnlessKept = [
  "__pycache__/",
  "*.pyc",
  ".venv/",
  "venv/",
  ".tox/",
  ".pytest_cache/",
  ".mypy_cache/",
  ".next/",
  ".nuxt/",
  ".svelte-kit/",
  ".turbo/",
  ".cache/",
  ".parcel-cache/",
  ".vercel/",
  ".netlify/",
  ".idea/",
  ".vscode/",
  "coverage/",
  "*.log",
]

/** Read in every folder, in this order. */
export const ignoreFileNames = [".gitignore", ".skifityignore"]

/** A .env with real values, as opposed to the template of one. */
export function isSecretsFile(name: string): boolean {
  if (name !== ".env" && !name.startsWith(".env.")) return false
  return ![".env.example", ".env.sample", ".env.template", ".env.dist"].includes(name)
}

type Rule = { base: string; pattern: RegExp; negate: boolean; dirOnly: boolean }

function parseRule(base: string, raw: string): Rule | null {
  let line = raw.replace(/[ \t\r]+$/, "")
  if (line === "" || line.startsWith("#")) return null
  let negate = false
  if (line.startsWith("!")) {
    negate = true
    line = line.slice(1)
  } else if (line.startsWith("\\!") || line.startsWith("\\#")) {
    line = line.slice(1)
  }
  let dirOnly = false
  if (line.endsWith("/")) {
    dirOnly = true
    line = line.replace(/\/+$/, "")
  }
  if (line === "") return null
  // A slash anywhere but the end ties the pattern to its own folder.
  const anchored = line.includes("/")
  line = line.replace(/^\//, "")
  const body = globToRegExp(line)
  try {
    const pattern = new RegExp(anchored ? `^${body}$` : `^(?:.*/)?${body}$`)
    return { base, pattern, negate, dirOnly }
  } catch {
    // A line git would not understand either; git skips it too.
    return null
  }
}

function escape(c: string): string {
  return c.replace(/[.*+?^${}()|[\]\\/-]/g, "\\$&")
}

function globToRegExp(glob: string): string {
  let out = ""
  for (let i = 0; i < glob.length; i++) {
    const rest = glob.slice(i)
    const c = glob[i]
    if (rest.startsWith("**/")) {
      out += "(?:.*/)?"
      i += 2
    } else if (rest.startsWith("/**") && i + 3 === glob.length) {
      out += "(?:/.*)?"
      i += 2
    } else if (rest.startsWith("**")) {
      out += ".*"
      i += 1
    } else if (c === "*") {
      out += "[^/]*"
    } else if (c === "?") {
      out += "[^/]"
    } else if (c === "[") {
      const end = glob.indexOf("]", i + 1)
      if (end < 0) {
        out += "\\["
        continue
      }
      let cls = glob.slice(i + 1, end)
      if (cls.startsWith("!")) cls = "^" + cls.slice(1)
      out += "[" + cls.replace(/\\/g, "\\\\") + "]"
      i = end
    } else if (c === "\\" && i + 1 < glob.length) {
      i++
      out += escape(glob[i])
    } else {
      out += escape(c)
    }
  }
  return out
}

function basename(path: string): string {
  return path.slice(path.lastIndexOf("/") + 1)
}

/**
 * Decides, path by path, what is sent.
 *
 * `ignoreFiles` maps a folder ("" for the top) to the text of its ignore files,
 * in the order of ignoreFileNames. A path is left out when it, or any folder it
 * is in, is — the same answer as the Go walk, which never enters a folder it
 * has left out, and so never reads that folder's own ignore files either.
 */
export function makeFolderFilter(ignoreFiles: Map<string, string[]>): (path: string) => boolean {
  const rules: Rule[] = []
  for (const line of leftOutUnlessKept) {
    const rule = parseRule("", line)
    if (rule) rules.push(rule)
  }
  const loaded = new Set<string>()
  const load = (dir: string) => {
    if (loaded.has(dir)) return
    loaded.add(dir)
    for (const text of ignoreFiles.get(dir) ?? []) {
      for (const line of text.split("\n")) {
        const rule = parseRule(dir, line)
        if (rule) rules.push(rule)
      }
    }
  }

  const ignored = (rel: string, isDir: boolean): boolean => {
    const name = basename(rel)
    if (alwaysLeftOut.has(name) || (!isDir && isSecretsFile(name))) return true
    let result = false
    for (const rule of rules) {
      if (rule.dirOnly && !isDir) continue
      let sub = rel
      if (rule.base !== "") {
        if (!rel.startsWith(rule.base + "/")) continue
        sub = rel.slice(rule.base.length + 1)
      }
      if (rule.pattern.test(sub)) result = !rule.negate
    }
    return result
  }

  const dirVerdicts = new Map<string, boolean>()
  load("")
  return (path: string) => {
    const parts = path.split("/")
    let dir = ""
    for (let i = 0; i < parts.length - 1; i++) {
      dir = dir === "" ? parts[i] : `${dir}/${parts[i]}`
      let out = dirVerdicts.get(dir)
      if (out === undefined) {
        out = ignored(dir, true)
        dirVerdicts.set(dir, out)
      }
      if (out) return false
      load(dir)
    }
    return !ignored(path, false)
  }
}

File: internal/cli/ignore.go
package cli

import (
	"bufio"
	"os"
	"path"
	"regexp"
	"strings"

	"skifity/internal/upload"
)

// What `skifity up` leaves out of the folder it sends.
//
// The folder on somebody's laptop is not the same thing as their app. It has
// node_modules in it, a .next cache, a virtualenv, the .env with their real
// keys. A repository already draws that line in .gitignore, and a folder an
// assistant generated almost always has one — so the same file draws it here,
// read the way git reads it. A .skifityignore beside it adds to it, for what is
// in git but should not be deployed.

// alwaysLeftOut is never sent, whatever an ignore file says: version control
// that is not source, dependencies the build installs itself, and the file this
// command writes for itself.
var alwaysLeftOut = map[string]bool{
	".git": true, ".hg": true, ".svn": true,
	"node_modules": true,
	".DS_Store":    true, "Thumbs.db": true,
	ProjectFileName: true,
}

// leftOutUnlessKept is left out as if a .gitignore said so, which means an
// ignore file can bring one back with a "!" line. Caches and environments that
// a build recreates, and editor folders.
var leftOutUnlessKept = []string{
	"__pycache__/", "*.pyc", ".venv/", "venv/", ".tox/", ".pytest_cache/", ".mypy_cache/",
	".next/", ".nuxt/", ".svelte-kit/", ".turbo/", ".cache/", ".parcel-cache/",
	".vercel/", ".netlify/", ".idea/", ".vscode/", "coverage/", "*.log",
}

// ignoreFiles are read in every directory, in this order.
var ignoreFiles = []string{".gitignore", ".skifityignore"}

type ignoreRule struct {
	// base is the directory the rule's file sits in, relative to the root. A
	// rule only applies beneath it.
	base    string
	pattern *regexp.Regexp
	negate  bool
	dirOnly bool
}

// ignorer decides, path by path, what stays behind.
type ignorer struct {
	rules []ignoreRule
}

func newIgnorer() *ignorer {
	ig := &ignorer{}
	for _, line := range leftOutUnlessKept {
		ig.add("", line)
	}
	return ig
}

// loadDir reads the ignore files in one directory. rel is the directory's path
// relative to the root, "" for the root itself.
func (ig *ignorer) loadDir(root, rel string) {
	for _, name := range ignoreFiles {
		file, err := os.Open(path.Join(root, rel, name))
		if err != nil {
			continue
		}
		scanner := bufio.NewScanner(file)
		for scanner.Scan() {
			ig.add(rel, scanner.Text())
		}
		file.Close()
	}
}

// add parses one line of an ignore file.
func (ig *ignorer) add(base, line string) {
	line = strings.TrimRight(line, " \t\r")
	if line == "" || strings.HasPrefix(line, "#") {
		return
	}
	rule := ignoreRule{base: base}
	if strings.HasPrefix(line, "!") {
		rule.negate = true
		line = line[1:]
	} else if strings.HasPrefix(line, `\!`) || strings.HasPrefix(line, `\#`) {
		line = line[1:]
	}
	if strings.HasSuffix(line, "/") {
		rule.dirOnly = true
		line = strings.TrimRight(line, "/")
	}
	if line == "" {
		return
	}
	// A slash anywhere but the end ties the pattern to the directory of the
	// file it is in. Without one it matches a name at any depth.
	anchored := strings.Contains(line, "/")
	line = strings.TrimPrefix(line, "/")
	expr := globToRegexp(line)
	if anchored {
		expr = "^" + expr + "$"
	} else {
		expr = "^(?:.*/)?" + expr + "$"
	}
	compiled, err := regexp.Compile(expr)
	if err != nil {
		// A line git would not understand either; skipping it is what git does.
		return
	}
	rule.pattern = compiled
	ig.rules = append(ig.rules, rule)
}

// ignored reports whether a path, relative to the root and with forward
// slashes, stays behind. The last rule that matches decides, as in git.
func (ig *ignorer) ignored(rel string, isDir bool) bool {
	name := path.Base(rel)
	if alwaysLeftOut[name] || (!isDir && upload.IsSecretsFile(name)) {
		return true
	}
	result := false
	for _, rule := range ig.rules {
		if rule.dirOnly && !isDir {
			continue
		}
		sub := rel
		if rule.base != "" {
			if !strings.HasPrefix(rel, rule.base+"/") {
				continue
			}
			sub = strings.TrimPrefix(rel, rule.base+"/")
		}
		if rule.pattern.MatchString(sub) {
			result = !rule.negate
		}
	}
	return result
}

// globToRegexp turns a gitignore glob into a regular expression body.
func globToRegexp(glob string) string {
	var b strings.Builder
	for i := 0; i < len(glob); i++ {
		c := glob[i]
		switch {
		case strings.HasPrefix(glob[i:], "**/"):
			b.WriteString("(?:.*/)?")
			i += 2
		case strings.HasPrefix(glob[i:], "/**") && i+3 == len(glob):
			b.WriteString("(?:/.*)?")
			i += 2
		case strings.HasPrefix(glob[i:], "**"):
			b.WriteString(".*")
			i++
		case c == '*':
			b.WriteString("[^/]*")
		case c == '?':
			b.WriteString("[^/]")
		case c == '[':
			end := strings.IndexByte(glob[i+1:], ']')
			if end < 0 {
				b.WriteString(`\[`)
				continue
			}
			class := glob[i+1 : i+1+end]
			if strings.HasPrefix(class, "!") {
				class = "^" + class[1:]
			}
			b.WriteString("[" + strings.ReplaceAll(class, `\`, `\\`) + "]")
			i += end + 1
		case c == '\\' && i+1 < len(glob):
			i++
			b.WriteString(regexp.QuoteMeta(string(glob[i])))
		default:
			b.WriteString(regexp.QuoteMeta(string(c)))
		}
	}
	return b.String()
}
